### PR TITLE
IIM-Intersection Code

### PR DESCRIPTION
**Pull request by ryanleaf1000**

### CODE REVIEW

File: include/fiddle/interaction/interaction_utilities.h [hunk 1]
```diff
@@ -238,5 +239,31 @@ namespace fdl
...
+   * WARNING: This code is specialized to the case in which q is a unit vector
+ aligned with the coordinate axes.
+   */
+  bool
+  intersect_line_with_edge(std::vector<std::pair<double, Point<1>> >& t_vals,
+                           const DoFHandler<1, 2>::active_cell_iterator elem,
+                           const dealii::MappingQ<1, 2> &mapping,
+                           dealii::Point<2> r,
+                           dealii::Tensor<1,2> q,
+                           const double tol = 0.0);
```
Comment (drwells) [inline include/fiddle/interaction/interaction_utilities.h:253]: While we may want to write two versions of this function for spacedim = 2 and spacedim = 3, that's generally an implementation detail - the general intersection methods should be templated on the dimension.

Other comments
- If there are preconditions then those should be assertions, not warnings listed in the documentation.
- Please use full words with underscores (or common abbreviations like `tria`) for input arguments
- Add doxygen-style documentation (like the other functions in this header)
- We should be using `Mapping` and not `MappingQ` (i.e., this won't work with simplices, and it definitely won't work with Lagrangian deformations via `MappingFEField`)

File: source/interaction/interaction_utilities.cc [hunk 2]
```diff
@@ -1132,6 +1133,387 @@ namespace fdl
...
+            // Solve for u:
+            //
+            //    a*u + b = 0
+            //
+            // with:
+            //
+            //    a = 0.5*(-p0+p1)
+            //    b = 0.5*(p0+p1) - r
+            const dealii::Point p0 = elem->vertex(0);
+            const dealii::Point p1 = elem->vertex(1);
```
Comment (drwells) [inline source/interaction/interaction_utilities.cc:1169]: Even though this works with C++17 we should still use `Point<spacedim>` here and below.

File: tests/grid/nodal_patch_map_multilevel_01.cc [hunk 3]
```diff
@@ -66,32 +66,26 @@ test(SAMRAI::tbox::Pointer<IBTK::AppInitializer> app_initializer)
                                                       {},
                                                       false,
                                                       partitioner);
-  // fudge the center a little so that things are not exactly on axes
-  Point<spacedim> center;
-  for (unsigned int d = 0; d < spacedim; ++d)
-    center[d] = 0.01;
-  GridGenerator::hyper_ball(tria, center);
+  GridGenerator::hyper_ball(tria);
```
Comment (drwells) [inline tests/grid/nodal_patch_map_multilevel_01.cc:69]: Why is this file changed?

**Review by drwells — CHANGES_REQUESTED** [review]

File: include/fiddle/interaction/interaction_utilities.h (hunk 1 above)
Comment (ryanleaf1000) [reply include/fiddle/interaction/interaction_utilities.h:253]: Thank you David. For the last comment, do you have any example or something similar related to judging what type of cell (whether it's simplices or quad) and type of mapping ?

Comment (drwells) [reply include/fiddle/interaction/interaction_utilities.h:253]: I recommend assuming a linear mapping for now - we can add some assertions and support for multiple intersections later. If you need the element reference cell you can always do `cell->reference_cell()` to distinguish between `Quadrilateral` and `Triangle` when `dim = 2` and `spacedim = 3`, though we shouldn't worry about quads yet.

Bear in mind that the position of the element is not defined by `cell->vertex(0)` etc: that's the reference configuration. The actual position is given by the mapping which uses a finite element field (`MappingFEField`) to compute the absolute positions.

Comment (drwells) [reply include/fiddle/interaction/interaction_utilities.h:253]: When we detect intersections we will need to read back Eulerian data - hence we should be using SAMRAI data here (in particular something like `SideIndex`) so that we don't lose information.

File: tests/grid/line_edge_intersection.cc [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+int main(int argc, char **argv)
+{
+  IBTK::IBTKInit ibtk_init(argc, argv, MPI_COMM_WORLD);
+  SAMRAI::tbox::Pointer<IBTK::AppInitializer> app_initializer =
+    new IBTK::AppInitializer(argc, argv, "multilevel_fe_01.log");
+
+      Triangulation<2> tria_bulk;
+      Triangulation<1,2> tria;
+      GridGenerator::hyper_ball(tria_bulk);
+      GridGenerator::hyper_sphere(tria,Point< 2 >(0,0), 1 );
```
Comment (drwells) [inline tests/grid/line_edge_intersection.cc:54]: I recommend using `GridGenerator::extract_boundary_mesh()` here and in the other test.

File: tests/grid/line_edge_intersection.cc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      tria.refine_global(2);
+
+      std::vector<BoundingBox<2>> cell_bboxes;
+      for (const auto &cell : tria.active_cell_iterators())
+        cell_bboxes.push_back(cell->bounding_box());
+      // Set up the relevant fiddle class:
```
Comment (drwells) [inline tests/grid/line_edge_intersection.cc:60]: Please don't add misleading comments there is no fiddle class used here). Also - why are you setting up bounding boxes?

File: tests/grid/line_edge_intersection.cc [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      std::vector<std::pair<double, Point<1>> > t_vals;
+      // now do the actual test
+      FE_Nothing<1,2> fe;
+      DoFHandler<1,2> dof_handler(tria);
+      dof_handler.distribute_dofs(fe);
+      for (const auto &cell : dof_handler.active_cell_iterators())
+        {
+          fdl::intersect_line_with_edge(t_vals,cell,mapping,r,q,-0.000);
+        }
+      return 1;
```
Comment (drwells) [inline tests/grid/line_edge_intersection.cc:76]: `return 1;` indicates an error: if we get here we should `return 0;` which is implicit in C++.

File: tests/grid/line_edge_intersection.cc [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+      std::vector<std::pair<double, Point<1>> > t_vals;
+      // now do the actual test
+      FE_Nothing<1,2> fe;
+      DoFHandler<1,2> dof_handler(tria);
+      dof_handler.distribute_dofs(fe);
+      for (const auto &cell : dof_handler.active_cell_iterators())
+        {
+          fdl::intersect_line_with_edge(t_vals,cell,mapping,r,q,-0.000);
+        }
```
Comment (drwells) [inline tests/grid/line_edge_intersection.cc:75]: This reveals an interesting problem with the API: it requires `DoFHandler::active_cell_iterator`s but doesn't actually use them as such. That should be reconsidered.

File: source/interaction/interaction_utilities.cc [hunk 8]
```diff
@@ -1132,6 +1133,387 @@ namespace fdl
...
+            if (q[0] != 0.0)
+              {
+                A1 = a[1];
+                A2 = a[2];
+                B1 = b[1];
+                B2 = b[2];
+                C1 = c[1];
+                C2 = c[2];
+                D1 = d[1] - r(1);
+                D2 = d[2] - r(2);
```
Comment (drwells) [inline source/interaction/interaction_utilities.cc:1323]: Can we rewrite this using `Tensor`? This code is excessively difficult to understand.

**Review by drwells — CHANGES_REQUESTED** [review]
I have a few more general comments:
1. You'll need to rebase since I did some header cleanup.
2. Tying back in to the original roadmap - we need a compressed format: e.g., tuples of SAMRAI indices describing the stencil and something else which describes where the intersection is. Using `t` as a convex combination parameter could work (we need a better name - no abbreviations please). For design reasons we need to batch-process all intersection points at once which should involve these utility functions.
3. Similarly, once we define a reasonable intermediate format for intersections, we can actually implement some IIM algorithms. We can do this concurrently - e.g., if we have flow past a block we can analytically compute the locations of the intersections right now to verify that everything else works. It might make sense to figure out which data structures would be useful for storing intersections before we actually write the intersection code.
4. Tests for fiddle are now automatically run in the CI (and can be run with `attest` in the same manner as IBAMR's tests). Check out how the other tests work and make sure that yours pass locally (and also in the CI).

File: source/interaction/interaction_utilities.cc [hunk 9]
```diff
@@ -1132,6 +1133,387 @@ namespace fdl
...
+                      if (v >= tol && (u + v) <= 1.0 + tol)
+                        {
+                          double t = f * (e2 * q);
+                          t_vals.push_back(std::make_pair(t, Point<2>(u, v)));
+                          std::cout << "line parameter t:"<<t<< std::endl;
+                        }
+                    }
+                }
+            };
+            for (unsigned int i = 0; i < pow(2,d_intersection_refinement-1); ++i)
```
[the remark this answers is not in the document]
Comment (ryanleaf1000) [reply source/interaction/interaction_utilities.cc:1453]: Yes. I agree with your comments. I felt even if we decide use this algorithm, we need some local refinement maybe base on curvatures. The reason I set d_intersection_refinement=5 is only a temporary parameter flag.

**Comment by drwells** [comment]
Can you get rid of the extra stuff in this pull request and limit it to just the intersection algorithms? Thanks!

**Comment by drwells** [comment]
More precisely - unlike the original version of the roadmap, I think it is wiser to just implement the basic intersection algorithms first and then handle the intersection-storing data structures concurrently. Lets focus on the intersection algorithms in this PR.

**Comment by ryanleaf1000** [comment]
Thank you for your feedback David, I didn't realize my update would also sync here. Please ignore all the unnecessary file and luck of documentation. This version just meant for my person save. I read the force spread part in current fiddle ifed, and I am trying to translate what we had in the previous IIM force spreading function to fiddle.

**Comment by drwells** [comment]
> Thank you for your feedback David, I didn't realize my update would also sync here. Please ignore all the unnecessary file and luck of documentation. This version just meant for my person save. 

Can you update this branch to just contain intersection algorithms? Furthermore - since we are going to be using linear elements for right now do you think it would be better to get rid of the `Mapping` arguments in the low-level intersection routines and instead just pass, e.g., `std::array<Point<spacedim>, spacedim>`?

> I read the force spread part in current fiddle ifed, and I am trying to translate what we had in the previous IIM force spreading function to fiddle.

The translation is going to require writing some intermediate data structures so be aware that whatever we do here will incorporate the iterator API from #184.

**Comment by ryanleaf1000** [comment]
Hello David, 
Sorry I was disctracted by students and forgot about the message,
Can you update this branch to just contain intersection algorithms?
I tried to made correction to exclude unnecessary files and changes. Do I also need to remove them from my online repos?
"Furthermore - since we are going to be using linear elements for right now do you think it would be better to get rid of the Mapping arguments in the low-level intersection routines and instead just pass, e.g., std::array<Point<spacedim>, spacedim>?"
Sorry I am not sure if I understand, are you suggesting we only consider linear elements for  intersection computation for now?
If so I totally agree without mapping is better.  But if you mean we use linear elements to search intersections for higher order element, I am not sure if we can get rid of the mapping.

**Comment by drwells** [comment]
> I tried to made correction to exclude unnecessary files and changes. Do I also need to remove them from my online repos?

Yes, lets remove them from this branch and force-push: we can deal with `IIMethod.h` etc. later.

> Sorry I am not sure if I understand, are you suggesting we only consider linear elements for intersection computation for now?

Yes - we should get that working before we move on to something more complex.

> But if you mean we use linear elements to search intersections for higher order element, I am not sure if we can get rid of the mapping.

No, we'll need to do something else in the higher-order case, but lets handle that once we have a second-order version of IIMethod working.

**Comment by drwells** [comment]
@ryanleaf1000 I think that we're actually not that far away from getting IIM working here - if we have these basic intersection algorithms then we have everything we need to tabulate all intersections (and then deduplicate). If we have intersections then we can use `FEPointEvaluation` to actually evaluate jump terms: the rest of the book-keeping is already implemented in `IFEDMethodBase`.

I'd like to get this done this semester and to achieve that goal I'd like to get all the intersection code (i.e., the basic linear intersections and `PatchIntersectionMap`) working by the end of the month. Do you think that's achievable? The clear next step for this PR is to get the low-level intersection routines working.

**Comment by ryanleaf1000** [comment]
Hello David,
I was working on fixing the IBAMR IIM implementation. And I focused there for last two weeks. I can switch back to here.
I tested the linear intersection code and I believe it is working properly. What would you suggest to do with PatchIntersectionMap? What do I need to add?

**Comment by ryanleaf1000** [comment]
Plus from our recent observation from current IIM tests. I think in order to make IIM working properly, higher order elements seems pretty necessary.

**Comment by drwells** [comment]
> I tested the linear intersection code and I believe it is working properly.

That's good to hear. There are few things that need to happen next with this PR:
- [ ] rebase
- [ ] write tests that are a part of the test suite
- [ ] remove all higher-order code and other unused files
- [ ] convert the given ones into strictly P1 functions: they take `std::array<Point<spacedim>, dim>` descriptions of elements in the deformed configuration and some description of the FD stencil and return some value which either indicates the coordinate along the array or nothing (one possibility would be to use `std::optional` here, available through deal.II as `std_cxx17::optional`). We can deal with the higher-order functions another day.

> I think in order to make IIM working properly, higher order representation of the geometry seems pretty necessary.

Could you expand on this? We've had good success so far with IIM with just linear elements (i.e., the first few papers).

> But probably it would be wise for us to also include FEMapping into our consideration.

Yes, but by writing things in layers it will be relatively easy to add that in later. Part of the reason I spent a lot of time on the roadmap was to make sure we can make that transition later.

**Comment by ryanleaf1000** [comment]
Ok. I can work on these this week.
"Could you expand on this? We've had good success so far with IIM with just linear elements (i.e., the first few papers)."
If you are interested we can discuss this in person.

**Comment by ryanleaf1000** [comment]
Good Morning David,
I just update this commit. I did 
 rebase
remove all higher-order code and other unused files
convert the given ones into strictly P1 functions
Is the current version of intersection functions look good to you? Or there are other changes you would prefer.

**Comment by ryanleaf1000** [comment]
Hello David,

I just fixed some errors. Do you want to talk about it? Or any suggested changes

**Comment by drwells** [comment]
You still need to rebase and get rid of the extra files.

**Comment by ryanleaf1000** [comment]
hello David, 
It's little bit too messy here. I just create another PR. Can we move there?

**Comment by drwells** [comment]
Why not rebase this one? Nevertheless we can move if you prefer.
